velox/functions/sparksql/String.h

```diff
@@ -1181,4 +1181,55 @@ struct FindInSetFunction {
   }
 };
 
+template <typename T>
+struct SoundexFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  static constexpr bool is_default_ascii_behavior = true;
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  void call(out_type<Varchar>& result, const arg_type<Varchar>& input) {
+    auto inputData = reinterpret_cast<const uint8_t*>(input.data());
+    size_t inputSize = input.size();
+    if (inputSize == 0) {
+      result.setEmpty();
+      return;
+    }
+    if (!std::isalpha(inputData[0])) {
+      // First character must be a letter, otherwise input is returned.
+      result.setNoCopy(input);
+      return;
+    }
+    result.resize(4);
+    result.data()[0] = std::toupper(inputData[0]);
+    int32_t sxi = 1;
+    int32_t idx = result.data()[0] - 'A';
+    char lastCode = kUSEnglishMapping[idx];
+    for (auto i = 1; i < inputSize; ++i) {
+      if (!std::isalpha(inputData[i])) {
+        lastCode = '0';
+        continue;
+      }
+      idx = std::toupper(inputData[i]) - 'A';
+      char code = kUSEnglishMapping[idx];
+      if (code != '7') {
```

Comment:
Can you make sure all the `if` branch is well tested?

Reply:
I think these are covered, please le me know if any missed. Thanks.